pilot/pkg/proxy/envoy/v2/ads.go

```diff
@@ -0,0 +1,394 @@
+}
+
+// XdsEvent represents a config or registry event that results in a push.
+type XdsEvent struct {
+
+	// If not empty, it is used to indicate the event is caused by a change in the clusters.
+	// Only EDS for the listed clusters will be sent.
+	clusters []string
+}
+
+// StreamAggregatedResources implements the ADS interface.
+func (s *DiscoveryServer) StreamAggregatedResources(stream ads.AggregatedDiscoveryService_StreamAggregatedResourcesServer) error {
+	peerInfo, ok := peer.FromContext(stream.Context())
+	peerAddr := "0.0.0.0"
+	if ok {
+		peerAddr = peerInfo.Addr.String()
+	}
+	var discReq *xdsapi.DiscoveryRequest
+	var receiveError error
+	reqChannel := make(chan *xdsapi.DiscoveryRequest, 1)
+
+	con := &XdsConnection{
+		pushChannel:   make(chan *XdsEvent, 1),
+		PeerAddr:      peerAddr,
+		Connect:       time.Now(),
+		HTTPListeners: []*xdsapi.Listener{},
+		RouteConfigs:  map[string][]*xdsapi.RouteConfiguration{},
+		Clusters:      []string{},
+		stream:        stream,
+	}
+	// Do not call: defer close(con.pushChannel) !
+	// the push channel will be garbage collected when the connection is no longer used.
+	// Closing the channel can cause subtle race conditions with push. According to the spec:
+	// "It's only necessary to close a channel when it is important to tell the receiving goroutines that all data
+	// have been sent."
+
+	// Reading from a stream is a blocking operation. Each connection needs to read
+	// discovery requests and wait for push commands on config change, so we add a
+	// go routine. If go grpc adds gochannel support for streams this will not be needed.
+	// This also detects close.
+	go func() {
+		defer close(reqChannel) // indicates close of the remote side.
+		for {
+			req, err := stream.Recv()
+			if err != nil {
+				if status.Code(err) == codes.Canceled || err == io.EOF {
+					log.Infof("ADS: %q %s terminated %v", peerAddr, con.ConID, err)
+					return
+				}
+				receiveError = err
+				log.Errorf("ADS: %q %s terminated with errors %v", peerAddr, con.ConID, err)
+				return
+			}
+			reqChannel <- req
+		}
+	}()
+	for {
+		// Block until either a request is received or the ticker ticks
+		select {
+		case discReq, ok = <-reqChannel:
+			if !ok {
+				// Remote side closed connection.
+				return receiveError
+			}
+			if discReq.Node.Id == "" {
+				log.Infof("Missing node id %s", discReq.String())
+				continue
+			}
+			nt, err := model.ParseServiceNode(discReq.Node.Id)
+			if err != nil {
+				return err
+			}
+			con.modelNode = &nt
+			if con.ConID == "" {
+				// first request
+				con.ConID = connectionID(discReq.Node.Id)
+			}
+
+			switch discReq.TypeUrl {
+			case ClusterType:
+				if con.CDSWatch {
+					// Already received a cluster watch request, this is an ACK
+					if discReq.ErrorDetail != nil {
+						log.Warnf("ADS:CDS: ACK ERROR %v %s %v", peerAddr, con.ConID, discReq.String())
+					}
+					if adsDebug {
+						log.Infof("ADS:CDS: ACK %v %v", peerAddr, discReq.String())
```

Comment:
Why not just use `log.Debugf` here?

Reply:
Because in most cases enabling debug results in an unreadable mess, and we don't enable it by default. 

This kind of info is very useful for debugging problems with XDS - since the new logging framework doesn't provide any way to enable debug by component that seems the best option.